### PR TITLE
Migrate from MCJIT to ORC JIT

### PR DESCRIPTION
**Pull request by dkurt**
resolves https://github.com/halide/Halide/issues/7161
resolves https://github.com/halide/Halide/issues/7078

* Migrate from MCJIT to ORC JIT
* Enables JIT on RISC-V (related: https://github.com/llvm/llvm-project/issues/58979)
* i386 is also on ORC JIT but fallbacks to RTDyld. Other targets uses JITLink. (track https://github.com/llvm/llvm-project/issues/58978)

### CODE REVIEW

File: src/JITModule.cpp [hunk 1]
```diff
@@ -262,30 +249,23 @@ void JITModule::compile_module(std::unique_ptr<llvm::Module> m, const string &fu
...
     string module_name = m->getModuleIdentifier();
 
-    llvm::EngineBuilder engine_builder((std::move(m)));
-    engine_builder.setTargetOptions(options);
-    engine_builder.setErrorStr(&error_string);
-    engine_builder.setEngineKind(llvm::EngineKind::JIT);
-    HalideJITMemoryManager *memory_manager = new HalideJITMemoryManager(dependencies);
-    engine_builder.setMCJITMemoryManager(std::unique_ptr<RTDyldMemoryManager>(memory_manager));
+    // Build TargetMachine
+    llvm::orc::JITTargetMachineBuilder tm_builder(llvm::Triple("x86_64-unknown-linux-gnu"));
```
Comment (dkurt) [inline src/JITModule.cpp:253]: Remove hardcoded triple

File: src/JITModule.cpp [hunk 2]
```diff
@@ -296,8 +276,39 @@ void JITModule::compile_module(std::unique_ptr<llvm::Module> m, const string &fu
     // TODO: If this ever works in LLVM, this would allow profiling of JIT code with symbols with oprofile.
     // listeners.push_back(llvm::createOProfileJITEventListener());
 
-    for (auto &listener : listeners) {
-        ee->RegisterJITEventListener(listener);
```
Comment (dkurt) [inline src/JITModule.cpp:300]: Need find a new place for it

File: src/JITModule.cpp [hunk 3]
```diff
@@ -296,9 +281,46 @@ void JITModule::compile_module(std::unique_ptr<llvm::Module> m, const string &fu
...
+    // Resolve symbol dependencies.
+    llvm::orc::SymbolMap NewSymbols;
+    auto symbolStringPool = JIT->getExecutionSession().getExecutorProcessControl().getSymbolStringPool();
+    for (const auto &module : dependencies) {
+        for (auto const &iter : module.exports()) {
+            orc::SymbolStringPtr name = symbolStringPool->intern(iter.first);
+            NewSymbols.insert({name, llvm::JITEvaluatedSymbol::fromPointer(iter.second.address)});
+        }
     }
+    llvm::cantFail(JIT->getMainJITDylib().define(llvm::orc::absoluteSymbols(std::move(NewSymbols))));
```
Comment (dkurt) [inline src/JITModule.cpp:320]: `addGenerator` works only for unresolved symbols. However there is a test `correctness_c_function` which replaces JIT extern. So we need redefine every dependency symbol (probably, in reversed order?)

File: src/JITModule.cpp [hunk 4]
```diff
@@ -296,9 +281,46 @@ void JITModule::compile_module(std::unique_ptr<llvm::Module> m, const string &fu
...
+    // Resolve system symbols (like pthread, dl and others)
+    JIT->getMainJITDylib().addGenerator(llvm::cantFail(llvm::orc::DynamicLibrarySearchGenerator::GetForCurrentProcess(target_data_layout.getGlobalPrefix())));
+
+    // Resolve symbol dependencies.
+    llvm::orc::SymbolMap NewSymbols;
+    auto symbolStringPool = JIT->getExecutionSession().getExecutorProcessControl().getSymbolStringPool();
+    for (const auto &module : dependencies) {
+        for (auto const &iter : module.exports()) {
+            orc::SymbolStringPtr name = symbolStringPool->intern(iter.first);
+            NewSymbols.insert({name, llvm::JITEvaluatedSymbol::fromPointer(iter.second.address)});
```
Comment (dkurt) [inline src/JITModule.cpp:317]: TODO: should we check that symbol already defined in previous dependency?

**Comment by dkurt** [comment]
Please review. All tests passed now (except [halide-testbranch-main-llvm16-arm-32-linux-cmake](https://buildbot.halide-lang.org/master/#/builders/100) which has not started today but used to do before).

File: src/JITModule.cpp [hunk 5]
```diff
@@ -262,43 +270,84 @@ void JITModule::compile_module(std::unique_ptr<llvm::Module> m, const string &fu
...
+                                  .setCompileFunctionCreator(compilerBuilder)
+                                  .setObjectLinkingLayerCreator(linkerBuilder)
+                                  .create());
+
+    auto ctors = llvm::orc::getConstructors(*m);
+    llvm::orc::CtorDtorRunner ctorRunner(JIT->getMainJITDylib());
+    ctorRunner.add(ctors);
+
+    auto dtors = llvm::orc::getDestructors(*m);
+    llvm::orc::CtorDtorRunner *dtorRunner = new llvm::orc::CtorDtorRunner(JIT->getMainJITDylib());
```
Comment (steven-johnson) [inline src/JITModule.cpp:323]: Should this be wrapped in `unique_ptr`? It's not clear to me where/how this gets disposed eventually.

File: src/JITModule.cpp [hunk 6]
```diff
@@ -262,43 +270,84 @@ void JITModule::compile_module(std::unique_ptr<llvm::Module> m, const string &fu
...
+    dtorRunner->add(dtors);
+
+    // Resolve system symbols (like pthread, dl and others)
+    JIT->getMainJITDylib().addGenerator(llvm::cantFail(llvm::orc::DynamicLibrarySearchGenerator::GetForCurrentProcess(target_data_layout.getGlobalPrefix())));
```
Comment (steven-johnson) [inline src/JITModule.cpp:327]: nit: this line is probably a bit too long, adding a break might be nice

File: src/JITModule.cpp [hunk 7]
```diff
@@ -262,43 +270,84 @@ void JITModule::compile_module(std::unique_ptr<llvm::Module> m, const string &fu
...
+
+    llvm::orc::ThreadSafeModule tsm(std::move(m), std::move(jit_module->context));
+    auto err = JIT->addIRModule(std::move(tsm));
+    internal_assert(!err) << llvm::toString(std::move(err)) << "\n";
+
+    // Resolve symbol dependencies
+    llvm::orc::SymbolMap NewSymbols;
```
Comment (steven-johnson) [inline src/JITModule.cpp:334]: nit: in Halide codebase, local vars should be initial-lowercase, `newSymbols`

File: src/JITModule.cpp [hunk 8]
```diff
@@ -310,31 +359,24 @@ void JITModule::compile_module(std::unique_ptr<llvm::Module> m, const string &fu
...
-    ee->finalizeObject();
-    // Do any target-specific post-compilation module meddling
-    for (auto &listener : listeners) {
-        ee->UnregisterJITEventListener(listener);
-        delete listener;
+        exports[requested_export] = compile_and_get_function(*JIT, requested_export);
     }
-    listeners.clear();
 
     // TODO: I don't think this is necessary, we shouldn't have any static constructors
```
Comment (steven-johnson) [inline src/JITModule.cpp:372]: Stale comment?

File: test/correctness/c_function.cpp [hunk 9]
```diff
@@ -71,7 +71,7 @@ int main(int argc, char **argv) {
     }
 
     if (call_counter2 != 32 * 32) {
-        printf("C function my_func2 was called %d times instead of %d\n", call_counter, 32 * 32);
+        printf("C function my_func2 was called %d times instead of %d\n", call_counter2, 32 * 32);
```
Comment (steven-johnson) [inline test/correctness/c_function.cpp:74]: eek, nice catch :-)

**Review by steven-johnson — COMMENTED** [review]
This looks pretty great! Let me pull it into Google's source and do some further testing, but I suspect this will be good once it passes our buildbots.

**Comment by steven-johnson** [comment]
Looks like a lot of failures on arm32, unfortunately ("JIT session error: Unsupported target machine architecture in ELF object shared runtime-jitted-objectbuffer")

**Comment by steven-johnson** [comment]
Google tests look good, just need to fix the arm32 cases

**Comment by dkurt** [comment]
@steven-johnson, thanks for comments! Switched 32 ARM to `RTDyld` similar to i386.

**Review by steven-johnson — APPROVED** [review]
LGTM, thanks for the fixes!

(Adding @abadams)

**Review by abadams — APPROVED** [review]

**Comment by steven-johnson** [comment]
(Will merge once all buildbots are green)

**Pull request merged**